File: algorithms/1710.maximum-units-on-a-truck/solution.rb
# @param {Integer[][]} box_types
# @param {Integer} truck_size
# @return {Integer}
def maximum_units(box_types, truck_size)
  res = 0

  box_types.sort { |a, b| b[1] <=> a[1] }.each do |num, unit|
    if num > truck_size
      res += truck_size * unit
      break
    else
      res += num * unit
      truck_size -= num
    end
  end

  res
end

File: algorithms/1710.maximum-units-on-a-truck/solution.ts
function maximumUnits(boxTypes: number[][], truckSize: number): number {
  const sortedBoxTypes = boxTypes.sort((a, b) => b[1] - a[1]);
  let res = 0;

  for (const v of sortedBoxTypes) {
    const [num, unit] = v;
    if (truckSize < num) {
      res += truckSize * unit;
      break;
    } else {
      res += num * unit;
      truckSize -= num;
    }
  }

  return res;
}
